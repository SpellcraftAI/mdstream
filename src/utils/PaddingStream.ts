export interface PaddingOptions {
  startPad?: number;
  endPad?: number;
  space?: string;
}

export class Padding {
  private currentLine: string = ""
  private firstChunk: boolean = true
  private lastLineLength: number = 0

  constructor(
    private startPad: number,
    private endPad: number,
    private space: string
  ) {}

  processChunk(text: string): string {
    let result = ""

    for (const char of text) {
      if (char !== "\n") {
        this.currentLine += char
        continue
      }

      const newline = `\n${this.space.repeat(this.startPad)}`
      const formattedLine = this.formatLine(
        this.currentLine,
        this.firstChunk ? this.startPad : 0,
        this.endPad
      )

      /**
       * Pad empty lines with spaces to match the length of the last line.
       */

      const emptyLine = this.currentLine.trim().length === 0
      result += formattedLine
      if (emptyLine) {
        result += this.space.repeat(this.lastLineLength)
      }
      result += newline

      this.lastLineLength = this.currentLine.length
      this.currentLine = ""
      this.firstChunk = false
    }

    return result
  }

  flush(): string {
    if (this.currentLine.length > 0) {
      return this.formatLine(this.currentLine, 0, this.endPad)
    }
    return ""
  }

  private formatLine(line: string, start: number = 0, end: number = 0): string {
    return this.space.repeat(start) + line + this.space.repeat(end)
  }
}

export class PaddingStream extends TransformStream<Uint8Array, Uint8Array> {
  private encoder = new TextEncoder()
  private decoder = new TextDecoder()

  constructor({ startPad = 1, endPad = 1, space = " " }: PaddingOptions = {}) {
    let paddingLogic: Padding

    super({
      start: () => {
        paddingLogic = new Padding(startPad, endPad, space)
      },
      transform: (chunk, controller) => {
        const text = this.decoder.decode(chunk, { stream: true })
        const processedText = paddingLogic.processChunk(text)
        if (processedText.length > 0) {
          controller.enqueue(this.encoder.encode(processedText))
        }
      },
      flush: (controller) => {
        const finalText = paddingLogic.flush()
        if (finalText.length > 0) {
          controller.enqueue(this.encoder.encode(finalText))
        }
      },
    })
  }
}