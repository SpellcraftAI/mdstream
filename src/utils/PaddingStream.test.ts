import { expect, test, describe } from "bun:test"
import { PaddingStream } from "./PaddingStream"

const ENCODER = new TextEncoder()
const DECODER = new TextDecoder()

describe("PaddingStream", () => {
  test("formats multi-line chunks correctly", async () => {
    const paddingStream = new PaddingStream({ space: "*", startPad: 2, endPad: 2 })
    const inputChunks = [
      "function hello() {\n",
      "  console.log('Hello,",
      " World!');\n",
      "  console.l",
      "og('abcxyz');\n",
      "}\n",
      "\n",
      "function bye() {}"
    ]

    let fullResult = ""
    const formattedChunks: string[] = []

    const readableStream = new ReadableStream<string>({
      start(controller) {
        for (const chunk of inputChunks) {
          controller.enqueue(chunk)
        }
        controller.close()
      }
    })

    const writableStream = new WritableStream<Uint8Array>({
      write(chunk) {
        const text = DECODER.decode(chunk)
        fullResult += text
        formattedChunks.push(text)
      }
    })

    await readableStream
      .pipeThrough(
        new TransformStream<string, Uint8Array>({
          transform: (chunk, controller) => {
            controller.enqueue(ENCODER.encode(chunk))
          },
        })
      )
      .pipeThrough(paddingStream)
      .pipeTo(writableStream)

    // Verify full result buffer
    expect(fullResult).toMatchSnapshot()
    console.log(fullResult)

    // Verify each chunk
    formattedChunks.forEach((chunk, index) => {
      expect(chunk).toMatchSnapshot(`Chunk ${index + 1}`)
    })
  })
})