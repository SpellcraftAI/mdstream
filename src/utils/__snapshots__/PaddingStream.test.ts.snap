// Bun Snapshot v1, https://goo.gl/fbAQLP

exports[`PaddingStream formats multi-line chunks correctly 1`] = `
"**function hello() {**
**  console.log('Hello, World!');**
**  console.log('abcxyz');**
**}**
*****
**function bye() {}**"
`;

exports[`PaddingStream formats multi-line chunks correctly: Chunk 1 1`] = `
"**function hello() {**
**"
`;

exports[`PaddingStream formats multi-line chunks correctly: Chunk 2 1`] = `
"  console.log('Hello, World!');**
**"
`;

exports[`PaddingStream formats multi-line chunks correctly: Chunk 3 1`] = `
"  console.log('abcxyz');**
**"
`;

exports[`PaddingStream formats multi-line chunks correctly: Chunk 4 1`] = `
"}**
**"
`;

exports[`PaddingStream formats multi-line chunks correctly: Chunk 5 1`] = `
"***
**"
`;

exports[`PaddingStream formats multi-line chunks correctly: Chunk 6 1`] = `"function bye() {}**"`;
